nncf/quantization/algorithms/hyperparameter_tuner/algorithm.py

```diff
@@ -0,0 +1,365 @@
+        :param validation_dataset: Dataset used to validate resulted model.
+        :param subset_indices: Zero-based indices of data items that should be selected
+            from the dataset and used to validate model.
+        :return: Resulted model.
+        """
+        self._set_backend_entity(model)
+
+        combinations = create_combinations(self._param_settings)
+
+        nncf_logger.info("Start initialization of algorithms")
+        with timer():
+            self._prepare_algorithms(model, combinations)
+
+        combination_score_fn = functools.partial(
+            self._calculate_combination_score,
+            initial_model=model,
+            dataset=validation_dataset,
+            subset_indices=subset_indices,
+        )
+
+        nncf_logger.info("Start search best combination of parameters")
+        with timer():
+            best_combination_key = find_best_combination(combinations, combination_score_fn, self._param_settings)
+
+        algorithm = self._algorithms[trim_zeros(best_combination_key)]
+        result_model = algorithm.apply(model, self._statistic_points)
+
+        return result_model
+
+    def _set_backend_entity(self, model: TModel) -> None:
+        """
+        Sets an entity with backend-specific logic of the algorithm.
+
+        :param model: Backend-specific model.
+        """
+        model_backend = get_backend(model)
+        if model_backend == BackendType.OPENVINO:
+            from nncf.quantization.algorithms.hyperparameter_tuner.openvino_backend import (
+                OVHyperparameterTunerAlgoBackend,
+            )
+
+            self._backend_entity = OVHyperparameterTunerAlgoBackend()
+        else:
+            raise RuntimeError(f"Cannot set backend-specific entity because {model_backend} is not supported!")
+
+    def _prepare_algorithms(self, initial_model: TModel, combinations: Dict[CombinationKey, Combination]) -> None:
+        """
+        Creates algorithm for each combination of parameters. Collects statistics for
+        created algorithms.
+
+        :param initial_model: Input model used to collect statistics for algorithms.
+        :param combinations: Combinations of parameters.
+        """
+        # Some combinations in `combinations` produce the same algorithm.
+        # More formally, two combination produce the same algorithm if their
+        # keys are equal after removing trailing zeros. So it is sufficient to
+        # consider only one such algorithm and discard the rest.
+        filtered_combinations = filter_combinations(combinations)
+
+        for trimed_combination_key, combination in filtered_combinations.items():
+            kwargs = apply_combination(self._init_params, combination)
+            self._algorithms[trimed_combination_key] = self._algorithm_cls(**kwargs)
+
+        # Collect required statistics for created algorithms
+        stats_aggregator = StatisticsAggregatorFactory.create(initial_model, self._calibration_dataset)
```

Comment:
Isn't it possible to collect statistics once, together with statistics for `MinMaxQuantization`?

Reply:
Yes, it is possible. I added TODO to resolve it in follow-up PR.